File: clients/python/pm_client.py
"""Minimal, dependency-light Python client + smoke test.

Copy this file into any sibling project - it is the whole integration.

    from pm_client import ProcessMiningClient

    pm = ProcessMiningClient("http://process-mining:8000", api_key="...")
    result = pm.mine_file("orders.csv", algorithm="dfg_performance", output_format="svg")
    open("map.svg", "w").write(result["result"]["image"])

Run it directly for an end-to-end smoke test:

    python pm_client.py --base-url http://localhost:8000 --api-key dev --file sample.csv
"""

from __future__ import annotations

import argparse
import json
import mimetypes
import urllib.error
import urllib.parse
import urllib.request
import uuid
from pathlib import Path
from typing import Any


class ProcessMiningError(RuntimeError):
    def __init__(self, status: int, body: Any) -> None:
        super().__init__(f"HTTP {status}: {body}")
        self.status = status
        self.body = body


class ProcessMiningClient:
    def __init__(self, base_url: str, api_key: str | None = None, timeout: int = 300) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    # ---- plumbing ------------------------------------------------------
    def _headers(self, extra: dict[str, str] | None = None) -> dict[str, str]:
        headers = {"Accept": "application/json"}
        if self.api_key:
            headers["X-API-Key"] = self.api_key
        headers.update(extra or {})
        return headers

    def _request(self, method: str, path: str, *, data: bytes | None = None,
                 headers: dict[str, str] | None = None, params: dict[str, Any] | None = None) -> Any:
        url = f"{self.base_url}{path}"
        if params:
            clean = {k: v for k, v in params.items() if v is not None}
            if clean:
                url += "?" + urllib.parse.urlencode(clean)
        request = urllib.request.Request(url, data=data, method=method, headers=self._headers(headers))
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                payload = response.read()
                content_type = response.headers.get("Content-Type", "")
                if "application/json" in content_type:
                    return json.loads(payload or b"null")
                return payload
        except urllib.error.HTTPError as exc:
            raw = exc.read()
            try:
                body = json.loads(raw)
            except json.JSONDecodeError:
                body = raw.decode("utf-8", errors="replace")
            raise ProcessMiningError(exc.code, body) from exc

    def _post_json(self, path: str, payload: dict[str, Any], **kwargs: Any) -> Any:
        return self._request(
            "POST",
            path,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"},
            **kwargs,
        )

    @staticmethod
    def _multipart(fields: dict[str, str], file_path: Path | None,
                   file_field: str = "file") -> tuple[bytes, str]:
        boundary = "----pm" + uuid.uuid4().hex
        chunks: list[bytes] = []
        for name, value in fields.items():
            if value is None:
                continue
            chunks.append(f"--{boundary}\r\n".encode())
            chunks.append(f'Content-Disposition: form-data; name="{name}"\r\n\r\n'.encode())
            chunks.append(f"{value}\r\n".encode())
        if file_path is not None:
            mime = mimetypes.guess_type(file_path.name)[0] or "application/octet-stream"
            chunks.append(f"--{boundary}\r\n".encode())
            chunks.append(
                f'Content-Disposition: form-data; name="{file_field}"; '
                f'filename="{file_path.name}"\r\n'.encode()
            )
            chunks.append(f"Content-Type: {mime}\r\n\r\n".encode())
            chunks.append(file_path.read_bytes())
            chunks.append(b"\r\n")
        chunks.append(f"--{boundary}--\r\n".encode())
        return b"".join(chunks), f"multipart/form-data; boundary={boundary}"

    # ---- API -----------------------------------------------------------
    def health(self) -> dict[str, Any]:
        return self._request("GET", "/health/ready")

    def profiles(self) -> list[dict[str, Any]]:
        return self._request("GET", "/api/v1/mapping-profiles")

    def mine_file(self, path: str | Path, *, algorithm: str = "dfg_frequency",
                  output_format: str = "json", mapping_profile: str | None = None,
                  columns: dict[str, Any] | None = None, filters: dict[str, Any] | None = None,
                  noise_threshold: float = 0.0, include_statistics: bool = True) -> dict[str, Any]:
        """Stateless: one call in, full answer out."""
        body, content_type = self._multipart(
            {
                "algorithm": algorithm,
                "format": output_format,
                "mapping_profile": mapping_profile,
                "columns": json.dumps(columns) if columns else None,
                "filters": json.dumps(filters) if filters else None,
                "noise_threshold": str(noise_threshold),
                "include_statistics": str(include_statistics).lower(),
            },
            Path(path),
        )
        return self._request("POST", "/api/v1/mine", data=body,
                             headers={"Content-Type": content_type})

    def upload_log(self, path: str | Path, *, name: str | None = None, tenant: str | None = None,
                   mapping_profile: str | None = None,
                   columns: dict[str, Any] | None = None) -> dict[str, Any]:
        body, content_type = self._multipart(
            {
                "name": name,
                "tenant": tenant,
                "mapping_profile": mapping_profile,
                "columns": json.dumps(columns) if columns else None,
            },
            Path(path),
        )
        return self._request("POST", "/api/v1/logs/upload", data=body,
                             headers={"Content-Type": content_type})

    def create_log(self, name: str, events: list[dict[str, Any]], **kwargs: Any) -> dict[str, Any]:
        return self._post_json("/api/v1/logs", {"name": name, "events": events, **kwargs})

    def append_events(self, log_id: str, events: list[dict[str, Any]]) -> dict[str, Any]:
        return self._post_json(f"/api/v1/logs/{log_id}/events", {"events": events})

    def discover(self, log_id: str, **payload: Any) -> dict[str, Any]:
        return self._post_json(f"/api/v1/logs/{log_id}/discover", payload)

    def process_map_svg(self, log_id: str, algorithm: str = "dfg_frequency") -> str:
        data = self._request("GET", f"/api/v1/logs/{log_id}/map",
                             params={"algorithm": algorithm, "format": "svg"})
        return data.decode("utf-8") if isinstance(data, bytes) else str(data)

    def statistics(self, log_id: str) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/logs/{log_id}/statistics")

    def variants(self, log_id: str, limit: int = 20) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/logs/{log_id}/variants", params={"limit": limit})

    def bottlenecks(self, log_id: str, limit: int = 10) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/logs/{log_id}/bottlenecks", params={"limit": limit})

    def conformance(self, log_id: str, **payload: Any) -> dict[str, Any]:
        return self._post_json(f"/api/v1/logs/{log_id}/conformance", payload)

    def delete_log(self, log_id: str) -> None:
        self._request("DELETE", f"/api/v1/logs/{log_id}")


def _smoke(base_url: str, api_key: str | None, file_path: str) -> None:
    client = ProcessMiningClient(base_url, api_key)
    print("health      :", client.health()["status"])
    print("profiles    :", [p["id"] for p in client.profiles()])

    uploaded = client.upload_log(file_path, name="smoke test", mapping_profile="bakery")
    log_id = uploaded["log"]["log_id"]
    print("log         :", log_id, uploaded["log"]["events"], "events")

    stats = client.statistics(log_id)
    print("cases       :", stats["cases"], "| median cycle:",
          round(stats["throughput_seconds"]["median"] / 60, 1), "min")

    top = client.bottlenecks(log_id, limit=3)["bottlenecks"]
    for item in top:
        print(f"bottleneck  : {item['source']} -> {item['target']} "
              f"{round(item['mean_duration_seconds'] / 60, 1)} min avg")

    svg = client.process_map_svg(log_id)
    Path("process_map.svg").write_text(svg, encoding="utf-8")
    print("svg         : process_map.svg", len(svg), "bytes")

    client.delete_log(log_id)
    print("cleanup     : ok")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Process Mining Service smoke test")
    parser.add_argument("--base-url", default="http://localhost:8000")
    parser.add_argument("--api-key", default=None)
    parser.add_argument("--file", default="examples/sample_log.csv")
    args = parser.parse_args()
    _smoke(args.base_url, args.api_key, args.file)

File: clients/js/pmClient.ts
/**
 * TypeScript client for the Process Mining Service.
 * Zero dependencies - works in Node 18+ and in the browser.
 */

export type Algorithm =
  | 'dfg_frequency'
  | 'dfg_performance'
  | 'petri_net_inductive'
  | 'petri_net_heuristics'
  | 'process_tree'
  | 'bpmn';

export type OutputFormat = 'json' | 'svg' | 'png' | 'dot';

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  frequency?: number | null;
  mean_duration_seconds?: number | null;
  metrics: Record<string, unknown>;
}

export interface GraphEdge {
  source: string;
  target: string;
  frequency?: number | null;
  mean_duration_seconds?: number | null;
  median_duration_seconds?: number | null;
  metrics: Record<string, unknown>;
}

export interface ProcessGraph {
  algorithm: Algorithm;
  nodes: GraphNode[];
  edges: GraphEdge[];
  start_activities: Record<string, number>;
  end_activities: Record<string, number>;
  stats: Record<string, unknown>;
}

export interface DiscoverResponse {
  log_id?: string | null;
  algorithm: Algorithm;
  format: string;
  cached: boolean;
  computed_in_ms: number;
  graph?: ProcessGraph | null;
  image?: string | null;
  content_type?: string | null;
}

export interface LogFilters {
  date_from?: string;
  date_to?: string;
  activities_include?: string[];
  activities_exclude?: string[];
  variant_coverage?: number;
  activity_coverage?: number;
  min_case_length?: number;
  max_case_length?: number;
}

export class ProcessMiningError extends Error {
  constructor(public status: number, public body: unknown) {
    super(`HTTP ${status}: ${JSON.stringify(body)}`);
  }
}

export class ProcessMiningClient {
  constructor(private baseUrl: string, private apiKey?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
  }

  private headers(extra: Record<string, string> = {}): Record<string, string> {
    return this.apiKey ? { 'X-API-Key': this.apiKey, ...extra } : extra;
  }

  private async handle<T>(response: Response): Promise<T> {
    const contentType = response.headers.get('content-type') || '';
    const body = contentType.includes('application/json')
      ? await response.json()
      : await response.text();
    if (!response.ok) throw new ProcessMiningError(response.status, body);
    return body as T;
  }

  /** Stateless: send a file, get the model back, nothing is stored server-side. */
  async mineFile(
    file: File | Blob,
    options: {
      algorithm?: Algorithm;
      format?: OutputFormat;
      mappingProfile?: string;
      columns?: Record<string, string>;
      filters?: LogFilters;
      includeStatistics?: boolean;
    } = {},
  ): Promise<{ result: DiscoverResponse; statistics?: unknown; bottlenecks?: unknown }> {
    const form = new FormData();
    form.append('file', file);
    form.append('algorithm', options.algorithm ?? 'dfg_frequency');
    form.append('format', options.format ?? 'json');
    form.append('include_statistics', String(options.includeStatistics ?? true));
    if (options.mappingProfile) form.append('mapping_profile', options.mappingProfile);
    if (options.columns) form.append('columns', JSON.stringify(options.columns));
    if (options.filters) form.append('filters', JSON.stringify(options.filters));

    const response = await fetch(`${this.baseUrl}/api/v1/mine`, {
      method: 'POST',
      headers: this.headers(),
      body: form,
    });
    return this.handle(response);
  }

  async uploadLog(file: File | Blob, options: { name?: string; tenant?: string; mappingProfile?: string } = {}) {
    const form = new FormData();
    form.append('file', file);
    if (options.name) form.append('name', options.name);
    if (options.tenant) form.append('tenant', options.tenant);
    if (options.mappingProfile) form.append('mapping_profile', options.mappingProfile);
    const response = await fetch(`${this.baseUrl}/api/v1/logs/upload`, {
      method: 'POST',
      headers: this.headers(),
      body: form,
    });
    return this.handle<{ log: Record<string, unknown>; warnings: string[] }>(response);
  }

  async appendEvents(logId: string, events: Array<Record<string, unknown>>) {
    const response = await fetch(`${this.baseUrl}/api/v1/logs/${logId}/events`, {
      method: 'POST',
      headers: this.headers({ 'Content-Type': 'application/json' }),
      body: JSON.stringify({ events }),
    });
    return this.handle(response);
  }

  async discover(logId: string, payload: Record<string, unknown> = {}): Promise<DiscoverResponse> {
    const response = await fetch(`${this.baseUrl}/api/v1/logs/${logId}/discover`, {
      method: 'POST',
      headers: this.headers({ 'Content-Type': 'application/json' }),
      body: JSON.stringify(payload),
    });
    return this.handle<DiscoverResponse>(response);
  }

  /** URL you can drop straight into an <img> / <object> tag. */
  mapUrl(logId: string, algorithm: Algorithm = 'dfg_frequency', format: OutputFormat = 'svg'): string {
    return `${this.baseUrl}/api/v1/logs/${logId}/map?algorithm=${algorithm}&format=${format}`;
  }

  async statistics(logId: string) {
    return this.handle(await fetch(`${this.baseUrl}/api/v1/logs/${logId}/statistics`, { headers: this.headers() }));
  }

  async variants(logId: string, limit = 20) {
    return this.handle(await fetch(`${this.baseUrl}/api/v1/logs/${logId}/variants?limit=${limit}`, { headers: this.headers() }));
  }

  async bottlenecks(logId: string, limit = 10) {
    return this.handle(await fetch(`${this.baseUrl}/api/v1/logs/${logId}/bottlenecks?limit=${limit}`, { headers: this.headers() }));
  }
}
